File: cpp/207.cpp
#include <iostream>
#include <queue>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
  bool canFinish(int numCourses, vector<vector<int>> &pre) {
    // init & build graph
    vector<vector<int>> g(numCourses, vector<int>());
    vector<int> in_deg(numCourses, 0);
    for (int i = 0; i < pre.size(); i++) {
      g[pre[i][1]].push_back(pre[i][0]);
      ++in_deg[pre[i][0]];
    }

    queue<int> deg0;
    for (int i = 0; i < numCourses; i++)
      if (!in_deg[i])
        deg0.push(i);

    vector<int> seq;
    while (deg0.size()) {
      int cur = deg0.front();
      seq.push_back(cur);
      deg0.pop();

      for (int i = 0; i < g[cur].size(); i++) {
        --in_deg[g[cur][i]];
        if (!in_deg[g[cur][i]])
          deg0.push(g[cur][i]);
      }
    }

    return seq.size() == numCourses;
  }
};

File: go/207.go
package main

func canFinish(numCourses int, prerequisites [][]int) bool {

	graph := make([][]int, numCourses)
	for i := range graph {
		graph[i] = make([]int, 0)
	}

	indeg := make([]int, numCourses)
	for i := 0; i < len(prerequisites); i++ {
		graph[prerequisites[i][1]] = append(graph[prerequisites[i][1]], prerequisites[i][0])
		indeg[prerequisites[i][0]]++
	}

	deg0 := make([]int, 0)
	for i := 0; i < numCourses; i++ {
		if indeg[i] == 0 {
			deg0 = append(deg0, i)
		}
	}

	seq := make([]int, 0)
	for len(deg0) > 0 {
		seq = append(seq, deg0[0])
		for i := 0; i < len(graph[deg0[0]]); i++ {
			indeg[graph[deg0[0]][i]]--
			if indeg[graph[deg0[0]][i]] == 0 {
				deg0 = append(deg0, graph[deg0[0]][i])
			}
		}
		deg0 = deg0[1:]
	}

	return len(seq) == numCourses
}
